### PR TITLE
[Android] `AsyncHydrationTrackerPlugin`

### PR DESCRIPTION
**Pull request by sugarmanz**
### Change Type (required)
Indicate the type of change your pull request is:

- [x] `patch`
- [ ] `minor`
- [ ] `major`

# Release Notes

Introduction of `AsyncHydrationTrackerPlugin` to provide a mechanism for reacting when `SuspendableAsset` hydration is completely finished.

```kotlin
androidPlayer.asyncHydrationTrackerPlugin!!.hooks.onHydrationComplete.tap(this::class.java.name) {
    // process effects after hydration is complete
}
```

### CODE REVIEW

File: android/player/src/main/java/com/intuit/player/android/asset/SuspendableAsset.kt [hunk 1]
```diff
@@ -50,13 +61,61 @@ public abstract class SuspendableAsset<Data>(assetContext: AssetContext, seriali
...
+
+    @ExperimentalPlayerApi
+    public class AsyncHydrationTrackerPlugin : AndroidPlayerPlugin {
+        private var trackedHydrations = mutableSetOf<String>()
+
+        public val hooks: Hooks = Hooks()
+
+        public fun SuspendableAsset<*>.trackHydration(): () -> Unit {
+            val key = assetContext.id
+            synchronized(trackedHydrations) {
```
Comment (brocollie08) [inline android/player/src/main/java/com/intuit/player/android/asset/SuspendableAsset.kt:88]: damn haven't seen this in a minute

File: android/player/src/main/java/com/intuit/player/android/asset/SuspendableAsset.kt [hunk 2]
```diff
@@ -50,13 +61,61 @@ public abstract class SuspendableAsset<Data>(assetContext: AssetContext, seriali
...
     private suspend fun View.doHydrate() = withContext(Dispatchers.Main) {
+        val onDone = track()
         try {
             hydrate(getData())
             setTag(R.bool.view_hydrated, true)
         } catch (exception: StaleViewException) {
             // b/c we're launched in a scope that isn't cared about anymore, we can't appropriately handle this, so just fast fail
             player.inProgressState?.fail(PlayerException("SuspendableAssets can't appropriately handle invalidateViews currently, this should be handled in a future major", exception))
+        } finally {
+            onDone()
```
Comment (brocollie08) [inline android/player/src/main/java/com/intuit/player/android/asset/SuspendableAsset.kt:76]: does this work? does this invocation not just return you the lambda?

Comment (brocollie08) [reply android/player/src/main/java/com/intuit/player/android/asset/SuspendableAsset.kt:76]: nvm, i think this is a case of let's not do this for clarity

**Review by brocollie08 — APPROVED** [review]

File: android/player/src/main/java/com/intuit/playerui/android/asset/SuspendableAsset.kt [hunk 3]
```diff
@@ -46,17 +57,71 @@ public abstract class SuspendableAsset<Data>(assetContext: AssetContext, seriali
     final override fun View.hydrate() {
         if (this is AsyncViewStub) return
 
+        player.asyncHydrationTrackerPlugin?.trackHydration(this@SuspendableAsset)
         setTag(R.bool.view_hydrated, false)
         hydrationScope.launch(Dispatchers.Main) { doHydrate() }
     }
 
     private suspend fun View.doHydrate() = withContext(Dispatchers.Main) {
+        player.asyncHydrationTrackerPlugin?.trackHydration(this@SuspendableAsset)
```
Comment (brocollie08) [inline android/player/src/main/java/com/intuit/playerui/android/asset/SuspendableAsset.kt:66]: what is the reason for calling this everywhere

Comment (sugarmanz) [reply android/player/src/main/java/com/intuit/playerui/android/asset/SuspendableAsset.kt:66]: There are several paths towards hydration:
- `initView` -> `doHydrate`
- `rehydrate` -> `hydrate`
- `initView` -> `hydrate`

We call this at each entry point to ensure that tracking is started as soon as possible, in a synchronous manner. I think the `doHydrate` on might be redundant. It was the original location, but we introduced to `initView` to ensure we started tracking hydration _before_ we launch the coroutine to create anything.

**Review by brocollie08 — APPROVED** [review]

**Pull request merged**